<a href="https://colab.research.google.com/github/michaelcassetti1/DATA301_Lab6/blob/main/DATA_301_Lab_6A_Michael_Cassetti.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# CityBikes Rest API

This section asks you to fetch JSON data from the [CityBikes REST API](https://api.citybik.es/v2/) to answer some questions about bike share programs.

Please be mindful not to send requests to the server more often than necessary. In particular, make sure you separate code that requests data from the server into a separate cell so that you do not have to repeatedly request data from the server. You will lose points if your requests are not separated into their own cell.

# Question 1

Find all cities with bike share programs in the United States (country code "US"), along with their network ID (for example, "ford-gobike"). How many cities in the U.S. have bike share programs (at least that are in this API)?

In [ ]:
import requests
import pandas as pd

json_data = requests.get("http://api.citybik.es/v2/networks").json()

cities = []
cities_id_pair = []

for network in json_data['networks']:
  if network['location']['country'] == "US":
    city = network['location']['city']
    network_id = network['id']
    cities.append(city)
    cities_id_pair.append((city, network_id))


print(len(cities_id_pair), "cities with bike share programs in the United States")
for city in cities_id_pair:
  print(city)

52 cities with bike share programs in the United States
('Aspen, CO', 'we-cycle')
('Ann Arbor, MI', 'arborbike')
('Austin, TX', 'austin')
('Chattanooga, TN', 'bike-chattanooga')
('Portland, OR', 'biketown')
('San Ramon, CA', 'britebikes')
('Boise, ID', 'boise-greenbike')
('Boulder, CO', 'boulder')
('Santa Monica, CA', 'breeze-bike-share')
('Fort Lauderdale, FL', 'broward')
('Milwaukee, WI', 'bublr-bikes')
('Buffalo, NY', 'reddy-bike-share')
('Washington, DC', 'capital-bikeshare')
('Charlotte, NC', 'charlotte')
('Cincinnati, OH', 'cincy-red-bike')
('New York, NY', 'citi-bike-nyc')
('Tampa, FL', 'coast-bike-share')
('Columbus, OH', 'cogo')
('Denver, CO', 'denver')
('Chicago, IL', 'divvy')
('El Paso, TX', 'elpaso')
('Fort Worth, TX', 'fortworth')
('Fargo, ND', 'greatrides')
('Salt Lake City', 'greenbikeslc')
('Phoenix, AZ', 'grid-bike-share')
('Omaha, NE', 'omaha')
('Houston, TX', 'houston')
('Boston, MA', 'blue-bikes')
('Philadelphia, PA', 'indego')
('Indianapolis, IN', 'indiana-pacers-b

# Question 2

Construct a `DataFrame` containing data about all bike stations in all networks in the United States. Save this `DataFrame` to disk using `.to_csv()`. (You will need it in Part B of this lab.)

Use this `DataFrame` to determine the total number of bicycles in bike share programs across the United States? You may assume that the number of bikes at a station is the number of empty spaces, plus the number of available bikes.

In [ ]:
import time
us_stations = []

for city in cities_id_pair:
  url = "http://api.citybik.es/v2/networks/" + city[1]
  json_data = requests.get(url).json()
  #us_stations.append(json_data)
  #time.sleep(0.5)
  #print(json_data['network']['stations'])
  for station in json_data['network']['stations']:
    city = city
    name = station['name']
    free_bikes = station['free_bikes']
    empty_slots = station['empty_slots']
    timestamp = station['timestamp']
    longitude = station['longitude']
    latitude = station['latitude']
    id = station['id']
    station_list = {'city': city,
                    'name': name,
                    'free_bikes': free_bikes,
                    'empty_slots': empty_slots,
                    'timestamp': timestamp,
                    'longitude': longitude,
                    'latitude': latitude,
                    'id': id,}
    us_stations.append(station_list)


df_stations = pd.DataFrame(us_stations)
df_stations




city                                       name  \
0       (Aspen, CO, we-cycle)                      Basalt BRT Downvalley   
1       (Aspen, CO, we-cycle)                              Triangle Park   
2       (Aspen, CO, we-cycle)                        Willits Town Center   
3       (Aspen, CO, we-cycle)                      El Jebel BRT Upvalley   
4       (Aspen, CO, we-cycle)                        JW Dr | El Jebel Rd   
...                       ...                                        ...   
8274  (Oklahoma, OK, spokies)  ST. OKC Blvd, North end of Park, entrance   
8275  (Oklahoma, OK, spokies)                    ST, S. Robinson and 5th   
8276  (Oklahoma, OK, spokies)                    ST, S. Robinson and 6th   
8277  (Oklahoma, OK, spokies)         UT, NW 23rd and Broadway, Byron's    
8278  (Oklahoma, OK, spokies)           CC, Dept of Agriculture and Food   

      free_bikes  empty_slots                    timestamp   longitude  \
0              0           13  2023-02-28T00:34:22.671000Z -107.034000   
1              0            7  2023-02-28T00:34:22.624000Z -107.083800   
2              0            7  2023-02-28T00:34:22.675000Z -107.083000   
3              0           11  2023-02-28T00:34:22.669000Z -107.089598   
4              0            7  2023-02-28T00:34:22.683000Z -107.090126   
...          ...          ...                          ...         ...   
8274           0            0  2023-02-28T00:34:36.963000Z  -97.517900   
8275           0            0  2023-02-28T00:34:36.964000Z  -97.516350   
8276           0            0  2023-02-28T00:34:36.965000Z  -97.516300   
8277           0            0  2023-02-28T00:34:36.911000Z  -97.485780   
8278           0            0  2023-02-28T00:34:36.936000Z  -97.501770   

       latitude                                id  
0     39.364081  2b0452ef99fa25ce9db62a8c9bb917aa  
1     39.387865  63cf805d373b399ae1daa391373d7a42  
2     39.386449  2ccb7b9041a6901cc76e92983e61cd9d  
3     39.392904  5fe55b6d158e753930aba287b87e0836  
4     39.399257  d4e64b89339be8aafa37203a16957568  
...         ...                               ...  
8274  35.462090  301c134c7aa65c3eaf7f0a8310ac7093  
8275  35.459880  df53692204f40d74d1988d7770252ba4  
8276  35.458990  ed4eace36b99ebf684d7c9f241a42f7b  
8277  35.531130  206c43d818ed14ffe19822ab9300c024  
8278  35.496950  1677e834802ea158ff273ef7653606df  

[8279 rows x 8 columns]

In [ ]:
df_stations['total_bikes'] = df_stations['free_bikes'] + df_stations['empty_slots']
df_stations

print(df_stations['total_bikes'].sum(), "total bikes in the US")

145565 total bikes in the US


In [ ]:
df_stations.to_csv('stations.csv', index = False)

# Question 3

You have just finished touring Coit Tower in San Francisco, which is located at latitude 37.802747 and longitude -122.405861. Using your `DataFrame` from Question 2, find the nearest bike station with an available bike, based on taxicab distance (a.k.a. Manhattan distance).

_Hint:_ You can check your answer using Google Maps!

In [ ]:
available_bike_stations = df_stations[df_stations['free_bikes'] > 0]
available_bike_stations['distance_from_Coit_Tower'] = (abs(37.802747 - available_bike_stations['latitude'])) + (abs((-122.405861 - available_bike_stations['longitude'])))
available_bike_stations = available_bike_stations.sort_values(by= 'distance_from_Coit_Tower')
available_bike_stations.iloc[0]


<ipython-input-68-64bcc8c25466>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  available_bike_stations['distance_from_Coit_Tower'] = (abs(37.802747 - available_bike_stations['latitude'])) + (abs((-122.405861 - available_bike_stations['longitude'])))


city                        (San Francisco Bay Area, CA, bay-wheels)
name                                   The Embarcadero at Sansome St
free_bikes                                                        12
empty_slots                                                       11
timestamp                                2023-02-28T00:33:41.053000Z
longitude                                                -122.403234
latitude                                                    37.80477
id                                  28754859653e0a656ba53e1fa0ea21db
total_bikes                                                       23
distance_from_Coit_Tower                                     0.00465
Name: 7648, dtype: object

## Submission Instructions

- Copy this notebook to your own Drive, if you have not already.
- Restart this notebook and run the cells from beginning to end. 
  - Go to Runtime > Restart and Run All.
- Rename this notebook by clicking on "DATA 301 Lab 6A - YOUR NAMES HERE" at the very top of this page. Replace "YOUR NAMES HERE" with the first and last names of you (and your partners, for Phase 2).
- Get the link to your notebook:
  - Click on "Share" at the top-right. 
  - Change the settings to "Anyone with the link can view". 
  - Copy the sharing link into Canvas.